code/Chapter5/test_mock_and_spy_template.cpp

```diff
@@ -0,0 +1,50 @@
+#include <iostream>
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+template <class GenericHorse>
+bool isAliveAtEndOfDay(int steps, double carbs, GenericHorse *animal)
+{
+    double spent_carbs{animal->walk(steps)};
+    if (spent_carbs > carbs)
+    {
+        animal->die();
+        return false;
+    }
+    animal->eat(carbs - spent_carbs);
+    return true;
+};
+
+class MockHorse
+{
+public:
+    MOCK_METHOD(double, walk, (int));
+    MOCK_METHOD(void, eat, (double));
+    MOCK_METHOD(void, die, ());
+};
+
+TEST(IsAliveTest, Lives)
+{
+    MockHorse animal = MockHorse();
```

Comment:
```suggestion
    MockHorse animal;
```
- I do not know why the original syntax should be preferred. You can directly create a `MockHorse` object. 
- This is also the reason you had to compile the code with C++17. With this change, it should work with lower standards also.
- The original statement would bring an additional overhead of assignment, for which it would need a copy constructor as well. However, after reading through a few sources, I am not sure if a Mock class has its default constructors or not. If it has, which ones are present and which ones are not? 
- In case you decide to implement the suggestion, you would need to make this change in all examples as this has been used at multiple places.
- Once your linker error goes away in the exercise code, the reason mentioned here might also help in figuring out further errors. You may need to define your own copy constructor (although not entirely sure about the last point).

Reply:
I'm the least expert in C++ here, but I understand that the preferred approach nowadays would be:

`MockHorse animal{MockHorse()}`

So I agree my code needs to be updated, but not to your option, but to this one, right?

Reply:
What you have shown is using a concept called uniform initialization y making use of constructor call explicitly. I do not think it is absolutely necessary to use uniform initialization here as no value is being passed.

So, we have three options in this case
1.   MockHorse animal;
2. MockHorse animal {}; // I think this should also work.
3. MockHorse animal{MockHorse()} 

In terms of performance, I do not know which one is better. My guess is all three should be same if no value is being used for initialization purpose.

Reply:
Option 1 calls the default constructor, as does option 2. Option 3 constructs an rvalue `MockHorse` (inside braces), and passes it to the move constructor of `MockHorse` (outside braces) which is unnecessary. I would go with 1 or 2. n.b.

As a side note, it's good to always use braces over parentheses when calling constructors. Else you might fall into the habit of doing e.g.  `MockHorse animal();`. is `animal` a variable or a function declaration? The compiler can't tell.

Reply:
Oh, scratch that, it seems the compiler is smart enough to work out that option 3 should just call default construction. See live example at [cpp.sh](https://cpp.sh/?source=%2F%2F+Example+program%0A%23include+%3Ciostream%3E%0A%23include+%3Cstring%3E%0A%0Aint+f(int+%26i)+%7B+return+--i%3B+%7D%0Aint+f(int+%26%26i)+%7B+return+%2B%2Bi%3B+%7D%0A%0Aint+main()%0A%7B%0A++++int+o+%3D+4%3B%0A++++int+%26lo+%3D+o%3B%0A++++int+%26%26ro+%3D+4%3B%0A%0A++++%2F%2Fint+%26%26ro1+%3D+o%3B%0A++++%2F%2Fint+%26%26ro2+%3D+lo%3B%0A++++%2F%2Fint+%26%26ro3+%3D+ro%3B%0A%0A++std%3A%3Acout+%3C%3C+%22Hello%2C+%22+%3C%3C+f(o)+%3C%3C+f(4)%3B%0A%7D). Though I would still say option 1 or 2 are more readable.

Reply:
Cool! I'll use option 1 and adapt the code, then. It's the simplest looking one. Many thanks for the explanations.

Reply:
Fixed in 11a6f04

To be honest, it is really confusing to have multiple ways of doing the same thing. I can understand reasons for backwards compatibility or performance, but it is really unhelpful not having and one only one way of doing things.